makefiles/dependency_resolution.inc.mk

```diff
@@ -0,0 +1,33 @@
+# Perform a recursive dependency resolution: Include $(RIOTBASE)/Makefile.dep
+# until no new modules and pkgs are pull in order to catch all transient
+# dependencies
+
+OLD_USEMODULE := $(sort $(USEMODULE))
+OLD_USEPKG:= $(sort $(USEPKG))
+
+# pull in dependencies of the currently used modules and pkgs
+include $(RIOTBASE)/Makefile.dep
+
+# check if required features are provided and update $(FEATURES_USED)
+include $(RIOTMAKE)/features_check.inc.mk
+
+# translate used features into used module, where needed
+include $(RIOTMAKE)/features_modules.inc.mk
+
+# sort and de-duplicate used modules and pkgs
+USEMODULE := $(sort $(USEMODULE))
+USEPKG := $(sort $(USEPKG))
+
+# If set of used modules and pkgs has changed during last run, run again to
+# recursively catch transitive dependencies
+ifneq ($(OLD_USEMODULE) $(OLD_USEPKG),$(USEMODULE) $(USEPKG))
```

Comment:
There are modules that are added depending on `FEATURES_USED`. See this example:
https://github.com/RIOT-OS/RIOT/blob/76acde098456ae6093b4cb269654362355b55429/cpu/cortexm_common/Makefile.dep#L16-L20

This relies on the fact that adding the feature to `FEATURES_OPTIONAL` immediately will update `FEATURES_USED` on next check, due to lazy evaluation. Given that you are changing the way `FEATURES_USED` is defined (with `:=`), what guarantees that there will be another iteration that will update `FEATURES_USED` and add the module? I think we should also keep an `OLD_FEATURES_USED` and trigger iterations based on that as well.

Reply:
True, I think this is also an issue in master though, there are other checks against `FEATURES_USED` with no guarantees of them being fulfiled:

- This is in `cpu/esp32/Makefile.dep` included first in `Makefile.dep`

```
ifneq (,$(filter cpp,$(FEATURES_USED)))
  USEMODULE += pthread
  USEMODULE += esp_cxx
endif
```
 
- this happens later in `Makefile.dep`

```
ifneq (,$(filter cpp11-compat,$(USEMODULE)))
  USEMODULE += xtimer
  USEMODULE += timex
  FEATURES_REQUIRED += cpp
endif

```

So the feature can become `REQUIRED` and then `USED` and the dependencies of the newly used `FEATURE` never be fulfilled.

This is actually causing a failure for some `periph` features as well:

```
ifneq (,$(filter esp_rtc_timer_32k,$(FEATURES_USED)))
    USEMODULE += esp_rtc_timer_32k
endif
```

Anyway I think @leandrolanzieri suggestion should fix these issues.

Reply:
> I think we should also keep an `OLD_FEATURES_USED` and trigger iterations based on that as well.

:+1: Will do!

A minimum example to show the issue you pointed out for the record:

``` Makefile
ifneq (,$(filter module_x,$(USEMODULE)))
  FEATURES_OPTIONAL += feature_a
  ifneq (,$(filter feature_a,$(FEATURES_USED)))
    USEMODULE += module_x_with_feature_a
  endif
endif

USEMODULE += module_x
```

After the fist iteration `module_x` is used, but `feature_a` is not yet marked as to be optionally used. After the second iteration `module_x` is still used and `feature_a` is marked as to be optionally used and will be added to `FEATURES_USED`. However, the third run needed for `module_x_with_feature_a` to be added is not done. (This would by the work, if `module_x` would be added before the `ifneq` expressions, making this hard to debug.)

Reply:
Done. Because the dependency state now consisting of modules, packages and features the line became long and difficult to read I added the variables `$(OLD_STATE)` and `$(NEW_STATE)`, which consist of the used modules, packages and features prior and after one iteration of the dependency resolution. (Also the sorting at the beginning is redundant, as the used stuff was sorted at the end of the iteration anyway. So I dropped this.)